File: MCN/routing/link-state.cpp
#include <iostream>
#include <climits>

using namespace std;

int main() {
    int n, src;
    cout << "Enter number of nodes: ";
    cin >> n;
    int INF = INT_MAX; 
    
    int cost[n][n], dist[n], visited[n];

    cout << "Enter cost matrix (use 9999 for no direct link):\n";
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> cost[i][j];
        }
        visited[i] = 0;
    }

    cout << "Enter source node (0 to " << n-1 << "): ";
    cin >> src;

    for (int i = 0; i < n; i++) {
        dist[i] = cost[src][i];
    }
    dist[src] = 0;
    visited[src] = 1;

    for (int count = 0; count < n - 1; count++) {
        int min = INF, u = -1;
        
        for (int i = 0; i < n; i++) {
            if (!visited[i] && dist[i] < min) {
                min = dist[i];
                u = i;
            }
        }
        
        if (u == -1) break; 
        visited[u] = 1;

        for (int v = 0; v < n; v++) {
            if (!visited[v] && cost[u][v] != INF && dist[u] + cost[u][v] < dist[v]) {
                dist[v] = dist[u] + cost[u][v];
            }
        }
    }

    cout << "\nRouting Table for Node " << src << ":\n";
    cout << "Destination\tShortest Distance\n";
    for (int i = 0; i < n; i++) {
        cout << i << "\t\t" << dist[i] << "\n";
    }

    return 0;
}

File: MCN py/routing/link-state.py
n = int(input("Enter number of nodes: "))
print("Enter cost matrix (row by row, use 9999 for infinity):")
cost = [list(map(int, input().split())) for _ in range(n)]
src = int(input("Enter source node: "))

dist = cost[src][:]
visited = [False] * n
visited[src] = True

for _ in range(n - 1):
    u = min((i for i in range(n) if not visited[i]), key=lambda i: dist[i])
    visited[u] = True
    
    for v in range(n):
        if not visited[v] and cost[u][v] != 9999:
            dist[v] = min(dist[v], dist[u] + cost[u][v])

print(f"Shortest distances from node {src}: {dist}")